### PR TITLE
CLN: pandas-dev#28926 Fix mypy errors in pandas/tests/io/parser/conftest.py

### PR DESCRIPTION
**Pull request by aaditya-panik**
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/io/parser/conftest.py [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
+from typing import List, Union
 
 import pytest
 
 from pandas import read_csv, read_table
 
 
 class BaseParser:
-    engine = None
+    engine = None  # type: Union[str, None]
```
Comment (WillAyd) [inline pandas/tests/io/parser/conftest.py]: Use `Optional[str]` instead of `Union[str, None]`

File: pandas/tests/io/parser/conftest.py [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
     low_memory = True
-    float_precision_choices = []
+    float_precision_choices = []  # type: List[Union[str, None]]
```
Comment (WillAyd) [inline pandas/tests/io/parser/conftest.py]: Same comment on Optional

File: pandas/tests/io/parser/conftest.py [hunk 3]
```diff
@@ -59,11 +60,11 @@ def csv1(csv_dir_path):
 
 _py_parsers_only = [_pythonParser]
 _c_parsers_only = [_cParserHighMemory, _cParserLowMemory]
-_all_parsers = _c_parsers_only + _py_parsers_only
+_all_parsers = [*_c_parsers_only, *_py_parsers_only]
```
Comment (WillAyd) [inline pandas/tests/io/parser/conftest.py]: Can you post the error this was causing? The addition seems easer to read so curious why we would need to unpack like this

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/io/parser/conftest.py (hunk 3 above)
Comment (aaditya-panik) [reply pandas/tests/io/parser/conftest.py]: As per the original issue, the error this gave was 
`pandas\tests\io\parser\conftest.py:62: error: Unsupported operand types for + ("List[CParser]" and "List[PythonParser]")`
I was confused if I should just type it as `List[BaseParser]` or the unpacking method. Unpacking is generally the prefered way according to PEP, but less readable as you rightly pointed out. Thinking back, I think the former is a better solution. Your thoughts?

Comment (WillAyd) [reply pandas/tests/io/parser/conftest.py]: Ah makes sense thanks for clarifying and linking PEP. I think this is OK but cc @simonjayhawkins who may have run into this before

Comment (aaditya-panik) [reply pandas/tests/io/parser/conftest.py]: Sure, @simonjayhawkins thoughts on this?

> As per the original issue, the error this gave was
> `pandas\tests\io\parser\conftest.py:62: error: Unsupported operand types for + ("List[CParser]" and "List[PythonParser]")`
> I was confused if I should just type it as `List[BaseParser]` or the unpacking method. Unpacking is generally the preferred way according to [PEP](https://www.python.org/dev/peps/pep-0448/), but less readable as you rightly pointed out. Thinking back, I think the former is a better solution. Your thoughts?

File: pandas/tests/io/parser/conftest.py (hunk 1 above)
Comment (aaditya-panik) [reply pandas/tests/io/parser/conftest.py]: Updated with bbb1fc6

File: pandas/tests/io/parser/conftest.py (hunk 2 above)
Comment (aaditya-panik) [reply pandas/tests/io/parser/conftest.py]: Updated with bbb1fc6

**Review by WillAyd — APPROVED** [review]
lgtm @simonjayhawkins

**Review by simonjayhawkins — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
Thanks @aaditya-panik

**Pull request merged**